### ISSUE TITLE
gambit-2.5 build issues

### ISSUE THREAD

**Issue by aseshkdatta** [GambitBSM/gambit_2.5#1]
Hi,

I am facing the following 'cmake' and 'make scanners' related issues while build gambit-2.5.
Could you please have a look and let me know how to circumvent these issues.

Thanks.
Asesh K Datta

========================================================================
On Ubuntu 22.04
-----------------------
cmake -DWITH_MPI=ON -DMPI_C_COMPILER=/home/asesh/Modules/Software/openmpi-4.1.6/install/bin/mpicc -DMPI_CXX_COMPILER=/home/asesh/Modules/Software/openmpi-4.1.6/install/bin/mpicxx -DMPI_Fortran_COMPILER=/home/asesh/Modules/Software/openmpi-4.1.6/install/bin/mpif90 -DWITH_AXEL=ON -DWITH_HEPMC=ON -DWITH_YODA=ON -DBUILD_FS_MODELS=MSSM -DWITH_RESTFRAMES=ON -DWITH_ROOT=ON  -DCMAKE_CXX_COMPILER=/usr/bin/g++ -DCMAKE_C_COMPILER=/usr/bin/gcc ..

** (Note that I haven't ditched "pybind11" as has been suggested to me earlier for  gambit-2.4.
    Should I continue  to ditch it?)

The issues are as follows:
-----------------------------------
**  No SQLite C libraries found. Excluding sqliteprinter and sqlitereader from GAMBIT configuration.
     Backends depending on SQLite3 (e.g. Contur) will be deactivated.

**     ColliderBit Solo (CBS) will be activated.
-- Performing Test CXX_SUPPORTS_no-unused-parameter
-- Performing Test CXX_SUPPORTS_no-unused-parameter - Failed
-- Performing Test CXX_SUPPORTS_no-deprecated-copy
-- Performing Test CXX_SUPPORTS_no-deprecated-copy - Failed
-- Performing Test CXX_SUPPORTS_no-sign-compare
-- Performing Test CXX_SUPPORTS_no-sign-compare - Failed

**  The versions of h5py (1.14.2) and hdf5 (1.10.7) do not match. Backends depending on h5py will be disabled

**  X Excluding darksusy 5.1.3 from GAMBIT configuration.
 X Excluding darksusy 6.1.1 from GAMBIT configuration.
 X Excluding darksusy 6.2.2 from GAMBIT configuration.
 X Excluding darksusy_MSSM 6.2.5 from GAMBIT configuration.
-- Performing Test CXX_SUPPORTS_no-format-overflow
-- Performing Test CXX_SUPPORTS_no-format-overflow - Failed
-- Performing Test CXX_SUPPORTS_no-format
-- Performing Test CXX_SUPPORTS_no-format - Failed
-- Performing Test CXX_SUPPORTS_no-implicit-function-declaration
-- Performing Test CXX_SUPPORTS_no-implicit-function-declaration - Failed
-- Performing Test CXX_SUPPORTS_no-type-limits
-- Performing Test CXX_SUPPORTS_no-type-limits - Failed
-- Performing Test CXX_SUPPORTS_no-maybe-uninitialized
-- Performing Test CXX_SUPPORTS_no-maybe-uninitialized - Failed
-- Performing Test CXX_SUPPORTS_no-unused-but-set-variable
-- Performing Test CXX_SUPPORTS_no-unused-but-set-variable - Failed
-- Performing Test CXX_SUPPORTS_no-extra
-- Performing Test CXX_SUPPORTS_no-extra - Failed
 X Excluding feynhiggs 2.11.2 from GAMBIT configuration.
 X Excluding feynhiggs 2.11.3 from GAMBIT configuration.
 X Excluding feynhiggs 2.12.0 from GAMBIT configuration.
-- Performing Test CXX_SUPPORTS_no-unused-local-typedefs
-- Performing Test CXX_SUPPORTS_no-unused-local-typedefs - Failed
-- Performing Test CXX_SUPPORTS_no-reorder
-- Performing Test CXX_SUPPORTS_no-reorder - Failed
-- Performing Test CXX_SUPPORTS_no-pessimizing-move
-- Performing Test CXX_SUPPORTS_no-pessimizing-move - Failed
-- Performing Test CXX_SUPPORTS_no-int-conversion
-- Performing Test CXX_SUPPORTS_no-int-conversion - Failed
-- Performing Test CXX_SUPPORTS_no-incompatible-pointer-types
-- Performing Test CXX_SUPPORTS_no-incompatible-pointer-types - Failed
-- Performing Test CXX_SUPPORTS_no-catch-value
-- Performing Test CXX_SUPPORTS_no-catch-value - Failed
-- Performing Test CXX_SUPPORTS_no-ignored-qualifiers
-- Performing Test CXX_SUPPORTS_no-ignored-qualifiers - Failed
-- Found Python module cython.
   Backends depending on Rivet's python extension will be enabled.

--------------------------------------------------------------------------------------------
make -j24 scanners
-------------
[ 34%] Building CXX object src/CMakeFiles/Minuit2.dir/MnMachinePrecision.cxx.o
[ 35%] Building CXX object src/CMakeFiles/Minuit2.dir/MnMinos.cxx.o
[ 36%] Building CXX object src/CMakeFiles/Minuit2.dir/MnParabolaFactory.cxx.o
In file included from /home/asesh/Packages/Gambit-BSM/gambit_2.5/ScannerBit/installed/minuit2/6.23.01/src/MPIProcess.cxx:11:
/home/asesh/Packages/Gambit-BSM/gambit_2.5/ScannerBit/installed/minuit2/6.23.01/inc/Minuit2/MPIProcess.h: In destructor ‘ROOT::Minuit2::MPITerminate::~MPITerminate()’:
/home/asesh/Packages/Gambit-BSM/gambit_2.5/ScannerBit/installed/minuit2/6.23.01/inc/Minuit2/MPIProcess.h:33:14: error: ‘MPI’ has not been declared
   33 |          if (MPI::Is_initialized() && !(MPI::Is_finalized())) {
      |              ^~~
/home/asesh/Packages/Gambit-BSM/gambit_2.5/ScannerBit/installed/minuit2/6.23.01/inc/Minuit2/MPIProcess.h:33:41: error: ‘MPI’ has not been declared
   33 |          if (MPI::Is_initialized() && !(MPI::Is_finalized())) {
      |                                         ^~~
/home/asesh/Packages/Gambit-BSM/gambit_2.5/ScannerBit/installed/minuit2/6.23.01/inc/Minuit2/MPIProcess.h:35:26: error: ‘MPI’ has not been declared
   35 |                       << MPI::COMM_WORLD.Get_rank() << " processor"
      |                          ^~~
/home/asesh/Packages/Gambit-BSM/gambit_2.5/ScannerBit/installed/minuit2/6.23.01/inc/Minuit2/MPIProcess.h:38:13: error: ‘MPI’ has not been declared
   38 |             MPI::Finalize();
      |             ^~~
/home/asesh/Packages/Gambit-BSM/gambit_2.5/ScannerBit/installed/minuit2/6.23.01/inc/Minuit2/MPIProcess.h: At global scope:
/home/asesh/Packages/Gambit-BSM/gambit_2.5/ScannerBit/installed/minuit2/6.23.01/inc/Minuit2/MPIProcess.h:133:14: error: ‘MPI’ does not name a type
  133 |       static MPI::Intracomm *fgCommunicator;
      |              ^~~
/home/asesh/Packages/Gambit-BSM/gambit_2.5/ScannerBit/installed/minuit2/6.23.01/inc/Minuit2/MPIProcess.h:135:14: error: ‘MPI’ does not name a type
  135 |       static MPI::Intracomm *fgCommunicators[2]; // maximum 2 communicators
      |              ^~~
/home/asesh/Packages/Gambit-BSM/gambit_2.5/ScannerBit/installed/minuit2/6.23.01/inc/Minuit2/MPIProcess.h: In static member function ‘static void ROOT::Minuit2::MPIProcess::StartMPI()’:
/home/asesh/Packages/Gambit-BSM/gambit_2.5/ScannerBit/installed/minuit2/6.23.01/inc/Minuit2/MPIProcess.h:74:16: error: ‘MPI’ has not been declared
   74 |          if (!(MPI::Is_initialized())) {
      |                ^~~
/home/asesh/Packages/Gambit-BSM/gambit_2.5/ScannerBit/installed/minuit2/6.23.01/inc/Minuit2/MPIProcess.h:75:13: error: ‘MPI’ has not been declared
   75 |             MPI::Init();
      |             ^~~
/home/asesh/Packages/Gambit-BSM/gambit_2.5/ScannerBit/installed/minuit2/6.23.01/inc/Minuit2/MPIProcess.h:77:26: error: ‘MPI’ has not been declared
   77 |                       << MPI::COMM_WORLD.Get_rank() << " processor"
      |                          ^~~
/home/asesh/Packages/Gambit-BSM/gambit_2.5/ScannerBit/installed/minuit2/6.23.01/inc/Minuit2/MPIProcess.h:80:25: error: ‘MPI’ has not been declared
   80 |          fgGlobalSize = MPI::COMM_WORLD.Get_size();
      |                         ^~~
/home/asesh/Packages/Gambit-BSM/gambit_2.5/ScannerBit/installed/minuit2/6.23.01/inc/Minuit2/MPIProcess.h:81:25: error: ‘MPI’ has not been declared
   81 |          fgGlobalRank = MPI::COMM_WORLD.Get_rank();
      |                         ^~~
/home/asesh/Packages/Gambit-BSM/gambit_2.5/ScannerBit/installed/minuit2/6.23.01/inc/Minuit2/MPIProcess.h: In static member function ‘static void ROOT::Minuit2::MPIProcess::TerminateMPI()’:
/home/asesh/Packages/Gambit-BSM/gambit_2.5/ScannerBit/installed/minuit2/6.23.01/inc/Minuit2/MPIProcess.h:87:14: error: ‘fgCommunicators’ was not declared in this scope
   87 |          if (fgCommunicators[0]!=0 && fgCommunicators[1]!=0) {
      |              ^~~~~~~~~~~~~~~
/home/asesh/Packages/Gambit-BSM/gambit_2.5/ScannerBit/installed/minuit2/6.23.01/inc/Minuit2/MPIProcess.h: In member function ‘void ROOT::Minuit2::MPIProcess::SumReduce(const double&, double&)’:
/home/asesh/Packages/Gambit-BSM/gambit_2.5/ScannerBit/installed/minuit2/6.23.01/inc/Minuit2/MPIProcess.h:105:13: error: ‘fgCommunicator’ was not declared in this scope
  105 |             fgCommunicator->Allreduce(&sub,&total,1,MPI::DOUBLE,MPI::SUM);
      |             ^~~~~~~~~~~~~~
/home/asesh/Packages/Gambit-BSM/gambit_2.5/ScannerBit/installed/minuit2/6.23.01/inc/Minuit2/MPIProcess.h:105:53: error: ‘MPI’ has not been declared
  105 |             fgCommunicator->Allreduce(&sub,&total,1,MPI::DOUBLE,MPI::SUM);
      |                                                     ^~~
/home/asesh/Packages/Gambit-BSM/gambit_2.5/ScannerBit/installed/minuit2/6.23.01/inc/Minuit2/MPIProcess.h:105:65: error: ‘MPI’ has not been declared
  105 |             fgCommunicator->Allreduce(&sub,&total,1,MPI::DOUBLE,MPI::SUM);
      |                                                                 ^~~
/home/asesh/Packages/Gambit-BSM/gambit_2.5/ScannerBit/installed/minuit2/6.23.01/src/MPIProcess.cxx: At global scope:
/home/asesh/Packages/Gambit-BSM/gambit_2.5/ScannerBit/installed/minuit2/6.23.01/src/MPIProcess.cxx
:::::::::::::::::::::::::::::::::::::::::::::::::::::::::
:::::::::::::::::::::::::::::::::::::::::::::::::::::::::
/home/asesh/Packages/Gambit-BSM/gambit_2.5/ScannerBit/installed/minuit2/6.23.01/inc/Minuit2/MPIProcess.h:105:13: error: ‘fgCommunicator’ was not declared in this scope
  105 |             fgCommunicator->Allreduce(&sub,&total,1,MPI::DOUBLE,MPI::SUM);
      |             ^~~~~~~~~~~~~~
/home/asesh/Packages/Gambit-BSM/gambit_2.5/ScannerBit/installed/minuit2/6.23.01/inc/Minuit2/MPIProcess.h:105:53: error: ‘MPI’ has not been declared
  105 |             fgCommunicator->Allreduce(&sub,&total,1,MPI::DOUBLE,MPI::SUM);
      |                                                     ^~~
/home/asesh/Packages/Gambit-BSM/gambit_2.5/ScannerBit/installed/minuit2/6.23.01/inc/Minuit2/MPIProcess.h:105:65: error: ‘MPI’ has not been declared
  105 |             fgCommunicator->Allreduce(&sub,&total,1,MPI::DOUBLE,MPI::SUM);
      |                                                                 ^~~
make[6]: *** [src/CMakeFiles/Minuit2.dir/build.make:496: src/CMakeFiles/Minuit2.dir/MnHesse.cxx.o] Error 1
make[5]: *** [CMakeFiles/Makefile2:193: src/CMakeFiles/Minuit2.dir/all] Error 2
make[4]: *** [Makefile:166: all] Error 2
make[3]: *** [CMakeFiles/minuit2_6.23.01.dir/build.make:88: minuit2_6.23.01-prefix/src/minuit2_6.23.01-stamp/minuit2_6.23.01-build] Error 2
make[2]: *** [CMakeFiles/Makefile2:4273: CMakeFiles/minuit2_6.23.01.dir/all] Error 2
make[2]: *** Waiting for unfinished jobs....
===========================================================

**Comment by tegonzalo** [comment]
Hi Asesh,

The pybind11 issue was fixed in this version, so you do not need to ditch it anymore.

As for this other issue, it does seem related to building minuit2 in your system. Replace the `make scanners` step of the build chain with `make diver polychord multinest`. That way you avoid building minuit2 which is causing the problem. 

Cheers,
Tomas

**Comment by aseshkdatta** [comment]
Thanks a lot, Tomas, for the suggestion. It worked.
However, gambit build exits with
-------------------------------------------
::::::::::::::::::::::::::::::::
::::::::::::::::::::::::::::::::
 /usr/bin/install -c -m 644 yoda-completion '/home/asesh/Packages/Gambit-BSM/gambit_2.5/contrib/YODA-1.9.7/local/etc/bash_completion.d'
Making install in pyext
Making install in yoda
Making install in .
 /usr/bin/mkdir -p '/home/asesh/Packages/Gambit-BSM/gambit_2.5/contrib/YODA-1.9.7/local/lib/pkgconfig'
 /usr/bin/install -c -m 644 yoda.pc '/home/asesh/Packages/Gambit-BSM/gambit_2.5/contrib/YODA-1.9.7/local/lib/pkgconfig'
[ 30%] Completed 'yoda'
[ 30%] Built target yoda
make[1]: *** [CMakeFiles/Makefile2:16134: CMakeFiles/gambit.dir/rule] Error 2
make: *** [Makefile:7430: gambit] Error 2
-------------------------------------------------------
Could you please observe?

Asesh

**Comment by tegonzalo** [comment]
There is indeed a build error there, but it must have happened earlier because the last lines before it just show YODA building successfully. Can you look at the build output earlier and try to see if something failed before? Either an error message or something like that.

**Comment by aseshkdatta** [comment]
Thanks, Tomas. Yes, I did find something like the following:
---------------------------------------------------------------------------------------------------------
In file included from ReaderYODA.cc:18:
../include/YODA/Profile2D.h: In member function 'YODA::Profile2D& YODA::Profile2D::operator=(const YODA::Profile2D&)':
../include/YODA/Profile2D.h:108:18: note: synthesized method 'YODA::Axis2D<YODA::ProfileBin2D, YODA::Dbn3D>& YODA::Axis2D<YODA::ProfileBin2D, YODA::Dbn3D>::operator=(const YODA::Axis2D<YODA::ProfileBin2D, YODA::Dbn3D>&)' first required here
  108 |       _axis = p2._axis;
      |                  ^~~~~
[ 15%] Linking CXX shared library ../../ScannerBit/lib/libscanner_multinest_3.11.so
In file included from /home/asesh/Packages/Gambit-BSM/gambit_2.5/ScannerBit/installed/great/1.0.0/MCMC/include/TGreatMCMCAlgorithm.h:9,
                 from /home/asesh/Packages/Gambit-BSM/gambit_2.5/ScannerBit/installed/great/1.0.0/MCMC/include/TGreatMCMCAlgorithmCovariance.h:4,
                 from /home/asesh/Packages/Gambit-BSM/gambit_2.5/ScannerBit/src/scanners/great/great.cpp:30:
/home/asesh/Packages/Gambit-BSM/gambit_2.5/ScannerBit/installed/great/1.0.0/Manager/include/TGreatEstimator.h: In member function ‘void TGreatEstimator<TAlgorithm>::ComputePDF()’:
/home/asesh/Packages/Gambit-BSM/gambit_2.5/ScannerBit/installed/great/1.0.0/Manager/include/TGreatEstimator.h:751:16: error: ISO C++ forbids variable length array ‘xmin’ [-Wvla]
  751 |         double xmin[fNPar];
      |                ^~~~
/home/asesh/Packages/Gambit-BSM/gambit_2.5/ScannerBit/installed/great/1.0.0/Manager/include/TGreatEstimator.h:752:16: error: ISO C++ forbids variable length array ‘xmax’ [-Wvla]
  752 |         double xmax[fNPar];
      |                ^~~~
/home/asesh/Packages/Gambit-BSM/gambit_2.5/ScannerBit/installed/great/1.0.0/Manager/include/TGreatEstimator.h: In member function ‘void TGreatEstimator<TAlgorithm>::ComputePDF(const int&, const char**, const char**)’:
/home/asesh/Packages/Gambit-BSM/gambit_2.5/ScannerBit/installed/great/1.0.0/Manager/include/TGreatEstimator.h:825:16: error: ISO C++ forbids variable length array ‘xmin’ [-Wvla]
  825 |         double xmin[nExp];
      |                ^~~~
/home/asesh/Packages/Gambit-BSM/gambit_2.5/ScannerBit/installed/great/1.0.0/Manager/include/TGreatEstimator.h:826:16: error: ISO C++ forbids variable length array ‘xmax’ [-Wvla]
  826 |         double xmax[nExp];
      |                ^~~~
/home/asesh/Packages/Gambit-BSM/gambit_2.5/ScannerBit/installed/great/1.0.0/Manager/include/TGreatEstimator.h:833:16: error: ISO C++ forbids variable length array ‘ParValues’ [-Wvla]
  833 |         double ParValues[gParameterSet->GetNParameters()];
      |                ^~~~~~~~~
/home/asesh/Packages/Gambit-BSM/gambit_2.5/ScannerBit/installed/great/1.0.0/Manager/include/TGreatEstimator.h:870:16: error: ISO C++ forbids variable length array ‘ExpValues’ [-Wvla]
  870 |         double ExpValues[nExp];
      |                ^~~~~~~~~
/home/asesh/Packages/Gambit-BSM/gambit_2.5/ScannerBit/installed/great/1.0.0/Manager/include/TGreatEstimator.h: In member function ‘int TGreatEstimator<TAlgorithm>::FillZone(Zone*)’:
/home/asesh/Packages/Gambit-BSM/gambit_2.5/ScannerBit/installed/great/1.0.0/Manager/include/TGreatEstimator.h:1221:16: error: ISO C++ forbids variable length array ‘x’ [-Wvla]
 1221 |         double x[fNPar];
      |                ^
/home/asesh/Packages/Gambit-BSM/gambit_2.5/ScannerBit/installed/great/1.0.0/Manager/include/TGreatEstimator.h: In member function ‘Zone* TGreatEstimator<TAlgorithm>::ComputeZone(const double&)’:
/home/asesh/Packages/Gambit-BSM/gambit_2.5/ScannerBit/installed/great/1.0.0/Manager/include/TGreatEstimator.h:1247:16: error: ISO C++ forbids variable length array ‘x’ [-Wvla]
 1247 |         double x[fNPar];
      |                ^
/home/asesh/Packages/Gambit-BSM/gambit_2.5/ScannerBit/installed/great/1.0.0/Manager/include/TGreatEstimator.h:1248:16: error: ISO C++ forbids variable length array ‘s’ [-Wvla]
 1248 |         double s[fNPar];
      |                ^
In file included from /home/asesh/Packages/Gambit-BSM/gambit_2.5/ScannerBit/installed/great/1.0.0/MCMC/include/TGreatMCMCAlgorithmCovariance.h:4,
------------------------------------------------------------------------------------------------
Kindly advise.
Asesh

**Comment by aseshkdatta** [comment]
Hi,

I dropped the following options at the 'cmake' stage

   HEPMC, YODA, RESTFRAMES, ROOT

and 'gambit' got built smoothly.

I am looking forward to use ColliderBit for which some of these are required, it seems.

Would you kindly look into the issue and let me know.

Cheers.
Asesh

**Comment by tegonzalo** [comment]
Uhm, it seems that the issue there is with `great`. Do you need it? `great` is just a scanner that uses root, but it seems that it hasn't been updated for some time and it clashes now with modern ISO standards. I'll report that to the Scanner working group so they can look into it. If you don't need `great` then you should just ditch it during cmake with `cmake -Ditch="great"`. 

As you say both HepMC and YODA are needed for ColliderBit, and they should build fine without any changes. Restframes and Root are only required if you are planning to use any of the recasted analyses that use them, e.g. ATLAS_13TeV_RJ3L_2Lep2Jets_36invfb. But if you are not planning to use them, maybe do not include Restframes or Root, with the cmake flags `-DWITH_RESTFRAMES=false` and `-DWITH_ROOT=false`

**Comment by aseshkdatta** [comment]
Thanks a lot, Tomas.

I have now ditched "great" and the build process completed smoothly.
In fact, I am interested in the recast so kept all of HepMC, YODA, Restframes and Root included.

However, while running "cmake" the following warning is sounded:
----------------------------------------------------------------------------------------
The versions of h5py (1.14.2) and hdf5 (1.10.7) do not match. Backends depending on h5py will be disabled
----------------------------------------------------------------------------------------

"./gambit backends" correctly shows the backend status (most of which are in red and are yet to be built).

However, when trying to check the backend requirements for "MSSM7.yaml" by executing

./gambit -bf yaml_files/MSSM7.yaml

it is Aborting by Dumping Core.
--------------------------------------------------------------------------------------------------
asesh@test-pc:~/Packages/Gambit-BSM/gambit_2.5$ ./gambit -bf yaml_files/MSSM7.yaml

GAMBIT 2.5.0
http://gambitbsm.org


RestFrames v1.0.0 -- Developed by Christopher Rogan (crogan@cern.ch)
                     Copyright (c) 2014-2018, Christopher Rogan
                     http://RestFrames.com

/home/asesh/.local/lib/python3.10/site-packages/h5py/__init__.py:36: UserWarning: h5py is running against HDF5 1.10.7 when it was built against 1.14.2, this may cause problems
  _warn(("h5py is running against HDF5 {0} when it was built against {1}, "
Warning! ***HDF5 library version mismatched error***
The HDF5 header files used to compile this application do not match
the version used by the HDF5 library to which this application is linked.
Data corruption or segmentation faults may occur if the application continues.
This can happen when an application was compiled by one version of HDF5 but
linked with a different version of static or shared HDF5 library.
You should recompile the application or check your shared library related
settings such as 'LD_LIBRARY_PATH'.
You can, at your own risk, disable this warning by setting the environment
variable 'HDF5_DISABLE_VERSION_CHECK' to a value of '1'.
Setting it to 2 or higher will suppress the warning messages totally.
Headers are 1.14.2, library is 1.10.7
	    SUMMARY OF THE HDF5 CONFIGURATION
	    =================================

General Information:
-------------------
                   HDF5 Version: 1.10.7
::::::::::::::::::::::::::::::::::::::::::::::::::::::::::::::::::::
::::::::::::::::::::::::::::::::::::::::::::::::::::::::::::::::::::
                        Fortran: yes
               Fortran Compiler: /usr/bin/gfortran
                  Fortran Flags: -g -O2 -ffile-prefix-map=$(top_srcdir)=. -flto=auto -ffat-lto-objects -fstack-protector-strong
               H5 Fortran Flags:  -std=f2008  -Waliasing -Wall -Wcharacter-truncation -Wextra -Wimplicit-interface -Wsurprising -Wunderflow -pedantic -Warray-temporaries -Wintrinsics-std -Wimplicit-procedure -Wreal-q-constant -Wfunction-elimination -Wrealloc-lhs -Wrealloc-lhs-all -Wno-c-binding-type -Wuse-without-only -Winteger-division -Wfrontend-loop-interchange  -s -O3
               AM Fortran Flags: 
         Shared Fortran Library: yes
         Static Fortran Library: yes

                            C++: yes
                   C++ Compiler: /usr/bin/g++
                      C++ Flags: -g -O2 -ffile-prefix-map=$(top_srcdir)=. -flto=auto -ffat-lto-objects -fstack-protector-strong -Wformat -Werror=format-security
                   H5 C++ Flags:   -Wall -Wcast-qual -Wconversion -Wctor-dtor-privacy -Weffc++ -Wextra -Wfloat-equal -Wformat=2 -Winit-self -Winvalid-pch -Wmissing-include-dirs -Wno-format-nonliteral -Wnon-virtual-dtor -Wold-style-cast -Woverloaded-virtual -Wreorder -Wshadow -Wsign-promo -Wundef -Wwrite-strings -pedantic -Wlarger-than=2560 -Wlogical-op -Wframe-larger-than=16384 -Wpacked-bitfield-compat -Wsync-nand -Wstrict-overflow=5 -Wno-unsuffixed-float-constants -Wdouble-promotion -Wtrampolines -Wstack-usage=8192 -Wmaybe-uninitialized -Wdate-time -Wopenmp-simd -Warray-bounds=2 -Wduplicated-cond -Whsa -Wnormalized -Wnull-dereference -Wunused-const-variable -Walloca -Walloc-zero -Wduplicated-branches -Wformat-overflow=2 -Wformat-truncation=1 -Wimplicit-fallthrough=5 -Wrestrict -Wattribute-alias -Wcast-align=strict -Wshift-overflow=2 -Wattribute-alias=2 -Wmissing-profile -fstdarg-opt -s -O3  -Wcast-align -Wmissing-declarations -Werror=packed -Werror=redundant-decls -Werror=switch -Wunused-but-set-variable -Werror=unused-function -Werror=unused-variable -Wunused-parameter -Werror=shadow
                   AM C++ Flags: 
             Shared C++ Library: yes
             Static C++ Library: yes

                           Java: yes
                  Java Compiler: /usr/bin/java (openjdk 11.0.13 2021-10-19)


Features:
---------
                   Parallel HDF5: no
Parallel Filtered Dataset Writes: no
              Large Parallel I/O: no
              High-level library: yes
                Build HDF5 Tests: yes
                Build HDF5 Tools: yes
                    Threadsafety: yes
             Default API mapping: v18
  With deprecated public symbols: yes
          I/O filters (external): deflate(zlib),szip(encoder)
                             MPE: no
                      Direct VFD: no
                      Mirror VFD: no
              (Read-Only) S3 VFD: yes
            (Read-Only) HDFS VFD: no
                         dmalloc: no
  Packages w/ extra debug output: none
                     API tracing: no
            Using memory checker: no
 Memory allocation sanity checks: no
          Function stack tracing: no
                Use file locking: best-effort
       Strict file format checks: no
    Optimization instrumentation: no
Bye...
Aborted (core dumped)
===================

Could you please observe/
Cheers,
Asesh

**Comment by aseshkdatta** [comment]
Hi Tomas,

I've now sorted out the above issue (versions conflict between hdf5 and h5py) by reinstalling h5py. Once, that is done, gambit got smoothly built.

However, as soon as I am trying to run gambit with the test file "MSSM7.yaml",
I'm facing issues that cause FATAL ERROR. The required backends for this test run
got smoothly built though, except in a few cases (like DarkSUSY_MSSM), a more recent version (v6.4.0) is built instead of the most recent version (v6.2.5) shown by
"/gambit -bf yaml_files/MSSM7.yaml". I am pasting its output below.

Could you please suggest a way out.

Cheers,
Asesh

(PS: I've had some smaller issues which I could sort out on my own but kept a list of:)
===============================================================
asesh@test-pc:~/Packages/Gambit-BSM/gambit_2.5$ ./gambit -bf yaml_files/MSSM7.yaml

GAMBIT 2.5.0
http://gambitbsm.org


RestFrames v1.0.0 -- Developed by Christopher Rogan (crogan@cern.ch)
                     Copyright (c) 2014-2018, Christopher Rogan
                     http://RestFrames.com

/home/asesh/.local/lib/python3.10/site-packages/numpy/core/getlimits.py:549: UserWarning: The value of the smallest subnormal for <class 'numpy.float64'> type is zero.
  setattr(self, word, getattr(machar, word).flat[0])
/home/asesh/.local/lib/python3.10/site-packages/numpy/core/getlimits.py:89: UserWarning: The value of the smallest subnormal for <class 'numpy.float64'> type is zero.
  return self._float_to_str(self.smallest_subnormal)
/home/asesh/.local/lib/python3.10/site-packages/numpy/core/getlimits.py:549: UserWarning: The value of the smallest subnormal for <class 'numpy.float32'> type is zero.
  setattr(self, word, getattr(machar, word).flat[0])
/home/asesh/.local/lib/python3.10/site-packages/numpy/core/getlimits.py:89: UserWarning: The value of the smallest subnormal for <class 'numpy.float32'> type is zero.
  return self._float_to_str(self.smallest_subnormal)

Starting GAMBIT
----------
Running in MPI-parallel mode with 1 processes
----------
Running with 32 OpenMP threads per MPI process (set by the environment variable OMP_NUM_THREADS).
YAML file: yaml_files/MSSM7.yaml
Importing: include/StandardModel_SLHA2_scan.yaml
Initialising logger...  log_debug_messages = true; log messages tagged as 'Debug' WILL be logged. 
WARNING: This may lead to very large log files!
Rank 0 resume flag? 1
Rank 0: output file readable? 1(filename: runs/MSSM7//samples//MSSM7.hdf5)
Resolving dependencies and backend requirements.  Hang tight...

Required backends to run file yaml_files/MSSM7.yaml
At least one backend candidate per row is required
--------------------------------------------------

(HiggsSignals, 1.4)      
(HiggsBounds, 4.2.1)     (HiggsBounds, 4.3.1)     
(HepLike, 2.0)           
(gamLike, 1.0.0)         (gamLike, 1.0.1)         
(DDCalc, 2.0.0)          
(DarkSUSY_MSSM, 6.4.0)   (DarkSUSY_MSSM, 6.1.1)   (DarkSUSY_MSSM, 6.2.5)   (DarkSUSY_MSSM, 6.2.2)   
(SuperIso, 4.1)          
(gm2calc, 1.3.0)         
(nulike, 1.0.8)          (nulike, 1.0.4)          (nulike, 1.0.7)          (nulike, 1.0.6)          (nulike, 1.0.5)          (nulike, 1.0.9)          
(SUSY_HIT, 1.5)          
(Pythia, 8.212)          

...done!
The scan you are about to run uses backends. Please make sure to cite all of them in your work.
You can find the list of references to include in GAMBIT.bib and an example TeX file in GAMBIT.tex


 FATAL ERROR

GAMBIT has exited with fatal exception: GAMBIT error
ERROR: A problem has been raised by the dependency resolver subsystem.
The following rules and options are not used in the current scan:

Module rules:

capability: DDCalc_Couplings
function: DDCalc_Couplings_WIMP_nucleon

Backend rules:

capability: DDCalc_Couplings
function: DDCalc_Couplings_WIMP_nucleon

Raised at: line 2176 in function void Gambit::DRes::DependencyResolver::checkForUnusedRules() of /home/asesh/Packages/Gambit-BSM/gambit_2.5/Core/src/depresolver.cpp.
Calling MPI_Finalize...
asesh@test-pc:~/Packages/Gambit-BSM/gambit_2.5$ 
=========================================

**Comment by aseshkdatta** [comment]
If I push through further and run gambit on the file "MSSM7.yaml" ignoring the dependency related issue, this is what I get (crazy fonts being displayed):
-------------------------------------------------------------------------------------------------------
./gambit -rf yaml_files/MSSM7.yaml 
(Following a suggestion from gambit, I had to insert the line 
   " delete_file_on_restart: true"  under the "Printer" block
in the "MSSM7.yaml" file to make gambit run! Don't think
I had to do this manually for the earlier gambit versions!)

:::::::::::::::::::::::::::::::::::::::::
:::::::::::::::::::::::::::::::::::::::::
    mZ: 91.1876
    theta12: 0.58376
    theta13: 0.15495
    theta23: 0.76958
  nuclear_params_sigmas_sigmal:
    deltad: -0.427
    deltas: -0.085
    deltau: 0.842
    sigmal: 31.9403
    sigmas: 47.8126
  Likelihood contribution from DarkBit::lnL_sigmas_sigmal: -10.4875
  Likelihood contribution from PrecisionBit::lnL_t_mass: -3.7347
  Likelihood contribution from PrecisionBit::lnL_mbmb: 2.58762
  Likelihood contribution from PrecisionBit::lnL_alpha_s: 5.05991
  Likelihood contribution from DarkBit::lnL_rho0_lognormal: -0.0310245
  Likelihood contribution from PrecisionBit::lnL_W_mass: -0.283052
  Likelihood contribution from PrecisionBit::lnL_gm2: 13.9472
�XY@V_�XY@V: Unknown observable type: �XY@V
  Likelihood contribution from FlavBit::HEPLike_b2sgamma_LogLikelihood: -244.942
�qd@V_�qd@V: Unknown observable type: �qd@V
  Likelihood contribution from FlavBit::HEPLike_B2Kstargamma_LogLikelihood: -603.781
��d@V_��d@V: Unknown observable type: ��d@V
d@V_d@V: Unknown observable type: d@V

 FATAL ERROR

GAMBIT has exited with fatal exception: map::at
Calling MPI_Finalize...

**Comment by ChrisJChang** [comment]
Hi Asesh,

I have only just been linked to this thread. Have you had any success since your last comment in June?

The dependency error about unused rules means that some of the rules in the yaml input are not actually used. In the past we have allowed these to stay in, but now we throw an error. Given that you also posted your next comment with a run having calculated, I assume you removed those unused rules from your input.

Regarding your last comment with the map::at error. This usually happens when there is a map somewhere and we are trying to access a non-existent key. I don't think I've ever seen the Unknown observable type error, with odd symbols. Could you look in your default log file (in your run/logs folder). I am guessing that something is messing up the name of an observable somewhere, and this might be stated in your log file as something like "Calling observable blah...", or "Printing blah".

**Comment by Dietr1ch** [comment]
I just tried to build and got similar failures, even after trying `-Ditch="great"` on the 2nd cmake invocation.

I'm trying to build with this `Justfile` (Dumb Makefile for times where `make` is busy building the recipes that `cmake` generated instead of the ones the User means to run).

```Justfile
build_gambit:
  @echo 'Building gambit'
  @echo '  Generating build instructions...'
  cmake -B build/ -S .  # Adding `-Ditch="great"` didn't help

  @echo '  building scanners...'
  make -j --directory build/ scanners

  @echo '  building gambit...'
  cmake -B build/ -S .  # Same, adding `-Ditch="great"` didn't help
  make -j --directory build/ gambit

build_backends: build_gambit
  @echo 'Building all backends'
  make -j --directory build/ backends

clean:
  @echo 'Removing build/'
  rm -rf ./build/
```


Running just (similar to make, it defaults the first target, `build_gambit`)
```bash
just
```

It fails on this command,
```bash
make -j --directory build/ scanners
```

With these errors,
```txt
make: Entering directory '~/Projects/gambit/build'
# ...
[ 75%] Performing configure step for 'polychord_1.20.1'

CMake Warning:
  Ignoring extra path from command line:

   "$HOME/Projects/gambit/ScannerBit/installed/minuit2"

CMake Warning:
  Ignoring extra path from command line:

   ".."
   


CMake Error: The current CMakeCache.txt directory $HOME/Projects/gambit/ScannerBit/installed/minuit2/6.23.01/CMakeCache.txt is different than the directory $HOME/Projects/gambit_2.5/ScannerBit/installed/minuit2/6.23.01 where CMakeCache.txt was created. This may result in binaries being created in the wrong place. If you are not sure, reedit the CMakeCache.txt
CMake Error: The source "$HOME/Projects/gambit/ScannerBit/installed/minuit2/6.23.01/CMakeLists.txt" does not match the source "$HOME/Projects/gambit_2.5/ScannerBit/installed/minuit2/6.23.01/CMakeLists.txt" used to generate cache.  Re-run cmake with a different source directory.
make[3]: *** [CMakeFiles/minuit2_6.23.01.dir/build.make:95: minuit2_6.23.01-prefix/src/minuit2_6.23.01-stamp/minuit2_6.23.01-configure] Error 1
make[2]: *** [CMakeFiles/Makefile2:3972: CMakeFiles/minuit2_6.23.01.dir/all] Error 2
make[2]: *** Waiting for unfinished jobs....

[100%] Performing configure step for 'multinest_3.12'
# ...

[100%] Built target multinest_3.12


make[1]: *** [CMakeFiles/Makefile2:2731: CMakeFiles/scanners.dir/rule] Error 2
make: *** [Makefile:839: scanners] Error 2
make: Leaving directory '$HOME/Projects/gambit/build'
error: Recipe `build_gambit` failed on line 11 with exit code 2
```


I'm using nix+devenv to set up [my environment](https://github.com/GambitBSM/gambit_2.5/compare/release_2.5...Dietr1ch:gambit:archive/2024-11-04/build/dev/nix-env), so this should be somewhat reproducible (Lacks the `-Ditch="great"` tweaks that didn't work :disappointed:).

**Comment by ChrisJChang** [comment]
Hi Dietr1ch,

This is not something I have seen before. I am suspicious of the fact that it is complaining about having a gambit folder and a different gambit_2.5 folder (both within the HOME/Projects folder). Are you trying to put your build folder outside of the main gambit source folder? In our builds we tend to put the build folder inside of the main gambit directory. Perhaps something in the minuit scanner build has assumed that was the case, so that there are differences not being correctly passed between your build folder and your main directory.

**Comment by Dietr1ch** [comment]
Yeah, I got bit by a directory move as I moved things to /tmp, I missed the mismatching paths, I thought the issue was just about the file contents.


I got things to build after adding a few more libraries. `zlib` wasn't mentioned as a dependency (which to be fair is probably already on every non-hermetic system).

Also, I also got bit by building with too many jobs (`-j` on 32 cores). I guess there's some race condition. I consistently got this obscure error until I limited the parallelism of `make gambit`. 16 cores did fine, but I'm not sure if can still randomly fail.
```txt
[ 25%] Built target yoda
make[1]: *** [CMakeFiles/Makefile2:15423: CMakeFiles/gambit.dir/rule] Error 2
make: *** [Makefile:7105: gambit] Error 2
```

**Comment by ChrisJChang** [comment]
Great. Yes, this race condition when building is a known issue that has been solved on some development branches, and will be merged into master in the near future.
